oauthproxy.go

```diff
@@ -89,10 +91,36 @@ func (u *UpstreamProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	u.handler.ServeHTTP(w, r)
 }
 
-func NewReverseProxy(target *url.URL) (proxy *httputil.ReverseProxy) {
-	return httputil.NewSingleHostReverseProxy(target)
+type ReverseProxy struct {
+	*httputil.ReverseProxy
+	wsProxy *websocketproxy.WebsocketProxy
 }
-func setProxyUpstreamHostHeader(proxy *httputil.ReverseProxy, target *url.URL) {
+
+func (p *ReverseProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if websocket.IsWebSocketUpgrade(req) {
+		p.wsProxy.ServeHTTP(rw, req)
+	} else {
+		p.ReverseProxy.ServeHTTP(rw, req)
+	}
+}
+
+func NewReverseProxy(target *url.URL) *ReverseProxy {
+	wsTarget := new(url.URL)
+	*wsTarget = *target
+	switch target.Scheme {
+	case "https":
+		wsTarget.Scheme = "wss"
+	case "http":
+		wsTarget.Scheme = "ws"
```

Comment:
this is a little weird. probably should change how the url resolution works, but they need the same handler so i'm going to punt.

Reply:
yes, https://www.w3.org/TR/websockets/#parsing-websocket-urls